README.md
## typedoc-plugin-internal-external

### What

A plugin for [Typedoc](http://typedoc.org)

Use an annotation (in a comment) to set code to "Internal" or "External".

### Examples

```
/** make this function @internal */
function internalFunction() {

}
```

```
/** make this class @external */
class ExternalClass {

}
```

```
/** @external Make all the code in this file (Dynamic Module) external */
/** A function in the dynamic module */
function func1() {
}
/** Another function in the dynamic module */
function func2() {
}
```


# Why

Typedoc categorizes your code into "Internal" and "External".
Essentially:

- Internal: Your own code
- External: Everything else

When using the default theme, typedoc provides a checkbox to show/hide the generated "External" documentation.

Typedoc uses the `files: []` array (in `tsconfig.json`) to determine if code is "Internal".
If a file being parsed is in the `files: []` array, then the code in that file is "Internal".

These annotations allow you to force code to be internal or external.

### Installing

Typedoc 0.4 has the ability to discover and load typedoc plugins found in node_modules.
Simply install the plugin and run typedoc.

```
npm install --save typedoc-plugin-internal-external
typedoc
```

### Using

Add `@internal` or `@external` to a comment.
That code's typedoc `Reflection` will have the `isExternal` boolean set accordingly.


```js
/**
 * @internal
 * This should always appear in the generated documentation
 */
class MyInternalClass {

}

/**
 * @external
 * This should only appear in the generated documentation when "Externals" is checked
 */
class MyExternalClass {

}
```


#### Annotation Aliases

Although the original purpose behind `Externals` was to hide documentation generated for external code,
you can use the show/hide feature of the default theme to hide whatever code you choose, by marking it as `@external`.
For example, you may have an internal API that you don't want shown in your docs by default.

Because marking "internal API" with "External" is counter-intuitive, you can choose an alias for the `@external`  annotation.

On the typedoc command line, define the aliases:

```
typedoc ....... --internal-aliases internal,publicapi --external-aliases external,internalapi
```

Then you can use those aliases in your comments:

```
/**
 * This should always appear in the generated documentation
 */
class PublicClass {

}

/**
 * @internalapi
 * This internal api's `Refletion` has `isExternal === true`, and should
 * only appear in the generated documentation when "Externals" is checked
 */
class InternalClass {

}
```


CHANGELOG.md
# 2.2.0 (2020-06-30)

[Compare `typedoc-plugin-internal-external` versions 2.1.1 and 2.2.0](https://github.com/christopherthielen/typedoc-plugin-internal-external/compare/2.1.1...2.2.0)

### Features

- Bring in backwards compatibility code from typedoc-plugin-external-module-name ([4452535](https://github.com/christopherthielen/typedoc-plugin-internal-external/commit/4452535))

## 2.1.1 (2020-01-19)

[Compare `typedoc-plugin-internal-external` versions 2.1.0 and 2.1.1](https://github.com/christopherthielen/typedoc-plugin-internal-external/compare/2.1.0...2.1.1)

### Bug Fixes

- typedoc 0.16 -- read options on EVENT_BEGIN, not initialize() ([0dd4c60](https://github.com/christopherthielen/typedoc-plugin-internal-external/commit/0dd4c60))

# 2.1.0 (2020-01-18)

[Compare `typedoc-plugin-internal-external` versions 2.0.1 and 2.1.0](https://github.com/christopherthielen/typedoc-plugin-internal-external/compare/2.0.1...2.1.0)

### Bug Fixes

- change test suite to start at 0.7.2. ([76053e5](https://github.com/christopherthielen/typedoc-plugin-internal-external/commit/76053e5))
- compatibility with typedoc 0.16.4 and higher ([0af0dd8](https://github.com/christopherthielen/typedoc-plugin-internal-external/commit/0af0dd8))
- fix for typedoc 0.16.x Options api ([f05d966](https://github.com/christopherthielen/typedoc-plugin-internal-external/commit/f05d966))


test/src/index.ts
/** @internal */
import './dir1/nest1';
import './external_module';
import './internal_module';
import './publicapi';
import './privateapi';
import './notspecified';

// export anything so typedoc >=0.16 will generate a module doc
export const str = 'string literal';


test/src/publicapi.ts
/** @publicapi a public module */
/** a public class */
export class PublicClass {}


test/src/dir1/nest1.ts
/**
 * a class
 * @internal
 */
export class Internal {}

/**
 * a class
 * @external
 */
export class External {}


plugin.ts
import { Reflection } from 'typedoc/dist/lib/models/reflections/abstract';
import { ReflectionKind } from 'typedoc/dist/lib/models/reflections';
import { Component, ConverterComponent } from 'typedoc/dist/lib/converter/components';
import { Converter } from 'typedoc/dist/lib/converter/converter';
import { Context } from 'typedoc/dist/lib/converter/context';
import { CommentPlugin } from 'typedoc/dist/lib/converter/plugins/CommentPlugin';
import { getRawComment } from 'typedoc/dist/lib/converter/factories/comment';
import { Options, OptionsReadMode } from 'typedoc/dist/lib/utils/options';

/**
 * This plugin allows you to specify if a symbol is internal or external.
 *
 * Add @internal or @external to the docs for a symbol.
 *
 * #### Example:
 * ```
 * &#47;**
 *  * @internal
 *  *&#47;
 * let foo = "123
 *
 * &#47;**
 *  * @external
 *  *&#47;
 * let bar = "123
 * ```
 */
@Component({ name: 'internal-external' })
export class InternalExternalPlugin extends ConverterComponent {
  externals: string[];
  internals: string[];

  externalRegex: RegExp;
  internalRegex: RegExp;

  initialize() {
    var options: Options = this.application.options;
    options.read({}, OptionsReadMode.Prefetch);

    this.externals = (options.getValue('external-aliases') || 'external').split(',');
    this.internals = (options.getValue('internal-aliases') || 'internal').split(',');

    this.externalRegex = new RegExp(`@(${this.externals.join('|')})\\b`);
    this.internalRegex = new RegExp(`@(${this.internals.join('|')})\\b`);

    this.listenTo(this.owner, {
      [Converter.EVENT_CREATE_SIGNATURE]: this.onSignature,
      [Converter.EVENT_CREATE_DECLARATION]: this.onDeclaration,
      [Converter.EVENT_FILE_BEGIN]: this.onFileBegin,
    });
  }

  private static markSignatureAndMethod(reflection: Reflection, external: boolean) {
    (reflection.flags as any).isExternal = external;
    // if (reflection.parent && (reflection.parent.kind === ReflectionKind.Method || reflection.parent.kind === ReflectionKind.Function) {
    if (reflection.parent && reflection.parent.kind & ReflectionKind.FunctionOrMethod) {
      (reflection.parent.flags as any).isExternal = external;
    }
  }

  /**
   * Triggered when the converter has created a declaration reflection.
   *
   * @param context  The context object describing the current state the converter is in.
   * @param reflection  The reflection that is currently processed.
   * @param node  The node that is currently processed if available.
   */
  private onSignature(context: Context, reflection: Reflection, node?) {
    if (!reflection.comment) return;

    // Look for @internal or @external
    let comment = reflection.comment;

    if (this.internals.some(tag => comment.hasTag(tag))) {
      InternalExternalPlugin.markSignatureAndMethod(reflection, false);
    } else if (this.externals.some(tag => comment.hasTag(tag))) {
      InternalExternalPlugin.markSignatureAndMethod(reflection, true);
    }

    this.internals.forEach(tag => CommentPlugin.removeTags(comment, tag));
    this.externals.forEach(tag => CommentPlugin.removeTags(comment, tag));
  }

  /**
   * Triggered when the converter has created a declaration reflection.
   *
   * @param context  The context object describing the current state the converter is in.
   * @param reflection  The reflection that is currently processed.
   * @param node  The node that is currently processed if available.
   */
  private onDeclaration(context: Context, reflection: Reflection, node?) {
    if (!reflection.comment) return;

    // Look for @internal or @external
    let comment = reflection.comment;

    if (this.internals.some(tag => comment.hasTag(tag))) {
      (reflection.flags as any).isExternal = false;
    } else if (this.externals.some(tag => comment.hasTag(tag))) {
      (reflection.flags as any).isExternal = true;
    }

    this.internals.forEach(tag => CommentPlugin.removeTags(comment, tag));
    this.externals.forEach(tag => CommentPlugin.removeTags(comment, tag));
  }

  /**
   * Triggered when the converter has started loading a file.
   *
   * This sets the file's context `isExternal` value if an annotation is found.
   * All symbols inside the file default to the file's `isExternal` value.
   *
   * The onFileBegin event is used because once the Declaration (which represents
   * the file) has been created, it's too late to update the context.
   * The declaration will also be processed during `onDeclaration` where the tags
   * will be removed from the comment.
   *
   * @param context  The context object describing the current state the converter is in.
   * @param reflection  The reflection that is currently processed.
   * @param node  The node that is currently processed if available.
   */
  private onFileBegin(context: Context, reflection: Reflection, node?) {
    if (!node) return;

    // Look for @internal or @external
    let comment = getRawComment(node);
    let internalMatch = this.internalRegex.exec(comment);
    let externalMatch = this.externalRegex.exec(comment);

    if (internalMatch) {
      context.isExternal = false;
    } else if (externalMatch) {
      context.isExternal = true;
    }
  }
}


typedocVersionCompatibility.ts
import * as ts from 'typescript';
import { Context } from 'typedoc/dist/lib/converter/context';
import { CommentPlugin } from 'typedoc/dist/lib/converter/plugins';
import { Comment, ProjectReflection } from 'typedoc/dist/lib/models';
import { memoize } from 'lodash';
import { satisfies } from 'semver';
import { Reflection } from 'typedoc/dist/lib/models/reflections';
import { ReflectionKind } from 'typedoc/dist/lib/models/reflections/abstract';
import { DeclarationReflection } from 'typedoc/dist/lib/models/reflections/declaration';

declare var require;
const typedocVersion = require('typedoc/package.json').version;

function checkTypedocVersion(semverString: string) {
  return satisfies(typedocVersion, semverString);
}

export const isTypedocVersion = memoize(checkTypedocVersion);

export function removeTags(comment: Comment, tag: string) {
  if (isTypedocVersion('< 0.17.0')) {
    return CommentPlugin.removeTags(comment, tag);
  } else {
    comment.removeTags(tag);
  }
}

export function removeReflection(project: ProjectReflection, reflection: Reflection) {
  if (isTypedocVersion('< 0.17.0')) {
    CommentPlugin.removeReflection(project, reflection);
  } else {
    project.removeReflection(reflection, true);
  }

  if (isTypedocVersion('>=0.16.0')) {
    delete project.reflections[reflection.id];
  }
}

export function createChildReflection(parent: Reflection, name: string) {
  if (isTypedocVersion('< 0.14.0')) {
    return new (DeclarationReflection as any)(parent, name, ReflectionKind.Module);
  } else {
    return new DeclarationReflection(name, ReflectionKind.Module, parent);
  }
}

/**
 * When we delete reflections, update the symbol mapping in order to fix:
 * https://github.com/christopherthielen/typedoc-plugin-external-module-name/issues/313
 * https://github.com/christopherthielen/typedoc-plugin-external-module-name/issues/193
 */
export function updateSymbolMapping(context: Context, symbol: ts.Symbol, reflection: Reflection) {
  if (!symbol) {
    return;
  }

  if (isTypedocVersion('< 0.16.0')) {
    // (context as any).registerReflection(reflection, null, symbol);
    (context.project as any).symbolMapping[(symbol as any).id] = reflection.id;
  } else {
    // context.registerReflection(reflection, symbol);
    const fqn = context.checker.getFullyQualifiedName(symbol);
    (context.project as any).fqnToReflectionIdMap.set(fqn, reflection.id);
  }
}

export function isModuleOrNamespace(reflection: Reflection) {
  if (isTypedocVersion('< 0.17.0')) {
    return reflection.kindOf((ReflectionKind as any).ExternalModule) || reflection.kindOf(ReflectionKind.Module);
  } else {
    return reflection.kindOf(ReflectionKind.Module) || reflection.kindOf(ReflectionKind.Namespace);
  }
}


typedoc-plugin-internal-external.ts
import { Component, ConverterComponent } from 'typedoc/dist/lib/converter/components';
import { Context } from 'typedoc/dist/lib/converter/context';
import { Converter } from 'typedoc/dist/lib/converter/converter';
import { ReflectionKind } from 'typedoc/dist/lib/models/reflections';
import { Reflection, ReflectionFlag, ReflectionFlags } from 'typedoc/dist/lib/models/reflections/abstract';
import { getRawComment } from './getRawComment';
import { removeTags } from './typedocVersionCompatibility';

function setExternal(flags: ReflectionFlags, isExternal: boolean) {
  if (typeof flags.setFlag === 'function') {
    flags.setFlag(ReflectionFlag.External, isExternal);
  } else {
    // probably not needed, but won't hurt
    (flags as any).isExternal = isExternal;
  }
}

/**
 * This plugin allows you to specify if a symbol is internal or external.
 *
 * Add @internal or @external to the docs for a symbol.
 *
 * #### Example:
 * ```
 * &#47;**
 *  * @internal
 *  *&#47;
 * let foo = "123
 *
 * &#47;**
 *  * @external
 *  *&#47;
 * let bar = "123
 * ```
 */
@Component({ name: 'internal-external' })
export class InternalExternalPlugin extends ConverterComponent {
  externals: string[];
  internals: string[];

  externalRegex: RegExp;
  internalRegex: RegExp;

  initialize() {
    this.listenTo(this.owner, {
      [Converter.EVENT_BEGIN]: this.readOptions,
      [Converter.EVENT_CREATE_SIGNATURE]: this.onSignature,
      [Converter.EVENT_CREATE_DECLARATION]: this.onDeclaration,
      [Converter.EVENT_FILE_BEGIN]: this.onFileBegin,
    });
  }

  private static markSignatureAndMethod(reflection: Reflection, external: boolean) {
    setExternal(reflection.flags, external);
    // if (reflection.parent && (reflection.parent.kind === ReflectionKind.Method || reflection.parent.kind === ReflectionKind.Function) {
    if (reflection.parent && reflection.parent.kind & ReflectionKind.FunctionOrMethod) {
      setExternal(reflection.parent.flags, external);
    }
  }

  private readOptions() {
    const { options } = this.application;

    this.externals = ((options.getValue('external-aliases') as string) || 'external').split(',');
    this.internals = ((options.getValue('internal-aliases') as string) || 'internal').split(',');

    this.externalRegex = new RegExp(`@(${this.externals.join('|')})\\b`);
    this.internalRegex = new RegExp(`@(${this.internals.join('|')})\\b`);
  }

  /**
   * Triggered when the converter has created a declaration reflection.
   *
   * @param context  The context object describing the current state the converter is in.
   * @param reflection  The reflection that is currently processed.
   * @param node  The node that is currently processed if available.
   */
  private onSignature(context: Context, reflection: Reflection, node?) {
    if (!reflection.comment) return;

    // Look for @internal or @external
    let comment = reflection.comment;

    if (this.internals.some((tag) => comment.hasTag(tag))) {
      InternalExternalPlugin.markSignatureAndMethod(reflection, false);
    } else if (this.externals.some((tag) => comment.hasTag(tag))) {
      InternalExternalPlugin.markSignatureAndMethod(reflection, true);
    }

    this.internals.forEach((tag) => removeTags(comment, tag));
    this.externals.forEach((tag) => removeTags(comment, tag));
  }

  /**
   * Triggered when the converter has created a declaration reflection.
   *
   * @param context  The context object describing the current state the converter is in.
   * @param reflection  The reflection that is currently processed.
   * @param node  The node that is currently processed if available.
   */
  private onDeclaration(context: Context, reflection: Reflection, node?) {
    if (!reflection.comment) return;

    // Look for @internal or @external
    let comment = reflection.comment;

    if (this.internals.some((tag) => comment.hasTag(tag))) {
      setExternal(reflection.flags, false);
    } else if (this.externals.some((tag) => comment.hasTag(tag))) {
      setExternal(reflection.flags, true);
    }

    this.internals.forEach((tag) => removeTags(comment, tag));
    this.externals.forEach((tag) => removeTags(comment, tag));
  }

  /**
   * Triggered when the converter has started loading a file.
   *
   * This sets the file's context `isExternal` value if an annotation is found.
   * All symbols inside the file default to the file's `isExternal` value.
   *
   * The onFileBegin event is used because once the Declaration (which represents
   * the file) has been created, it's too late to update the context.
   * The declaration will also be processed during `onDeclaration` where the tags
   * will be removed from the comment.
   *
   * @param context  The context object describing the current state the converter is in.
   * @param reflection  The reflection that is currently processed.
   * @param node  The node that is currently processed if available.
   */
  private onFileBegin(context: Context, reflection: Reflection, node?) {
    if (!node) return;

    // Look for @internal or @external
    let comment = getRawComment(node);
    let internalMatch = this.internalRegex.exec(comment);
    let externalMatch = this.externalRegex.exec(comment);

    if (internalMatch) {
      context.isExternal = false;
    } else if (externalMatch) {
      context.isExternal = true;
    }
  }
}
